File: goldenmatch/core/cluster.py
"""Union-Find clustering for GoldenMatch."""

from __future__ import annotations

from collections import defaultdict


class UnionFind:
    """Union-Find (disjoint set) with path compression and union by rank."""

    def __init__(self) -> None:
        self._parent: dict[int, int] = {}
        self._rank: dict[int, int] = {}

    def add(self, x: int) -> None:
        """Add an element as its own root."""
        if x not in self._parent:
            self._parent[x] = x
            self._rank[x] = 0

    def add_many(self, ids: list[int]) -> None:
        """Add multiple elements at once, more efficient than individual add() calls."""
        parent = self._parent
        rank = self._rank
        for x in ids:
            if x not in parent:
                parent[x] = x
                rank[x] = 0

    def find(self, x: int) -> int:
        """Find the root of x with iterative path compression."""
        root = x
        while self._parent[root] != root:
            root = self._parent[root]
        # Path compression: point all nodes on the path directly to root
        while self._parent[x] != root:
            next_x = self._parent[x]
            self._parent[x] = root
            x = next_x
        return root

    def union(self, a: int, b: int) -> None:
        """Union the sets containing a and b using union by rank."""
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return
        if self._rank[ra] < self._rank[rb]:
            ra, rb = rb, ra
        self._parent[rb] = ra
        if self._rank[ra] == self._rank[rb]:
            self._rank[ra] += 1

    def get_clusters(self) -> list[set[int]]:
        """Return all clusters as a list of sets."""
        groups: dict[int, set[int]] = defaultdict(set)
        for x in self._parent:
            groups[self.find(x)].add(x)
        return list(groups.values())


def _build_mst(
    members: list[int], pair_scores: dict[tuple[int, int], float],
) -> list[tuple[int, int, float]]:
    """Build max-weight spanning tree using Kruskal's algorithm."""
    edges = [(a, b, s) for (a, b), s in pair_scores.items()]
    edges.sort(key=lambda e: e[2], reverse=True)
    uf = UnionFind()
    uf.add_many(members)
    mst: list[tuple[int, int, float]] = []
    for a, b, s in edges:
        if uf.find(a) != uf.find(b):
            uf.union(a, b)
            mst.append((a, b, s))
            if len(mst) == len(members) - 1:
                break
    return mst


def split_oversized_cluster(
    members: list[int], pair_scores: dict[tuple[int, int], float],
) -> list[dict]:
    """Split a cluster by removing the weakest MST edge."""
    if len(members) <= 1 or not pair_scores:
        return [{"members": sorted(members), "size": len(members),
                 "oversized": False, "pair_scores": pair_scores}]

    mst = _build_mst(members, pair_scores)
    if not mst:
        return [{"members": sorted(members), "size": len(members),
                 "oversized": False, "pair_scores": pair_scores}]

    weakest = min(mst, key=lambda e: e[2])
    remaining = [(a, b, s) for a, b, s in mst if (a, b, s) != weakest]

    uf = UnionFind()
    uf.add_many(members)
    for a, b, _s in remaining:
        uf.union(a, b)

    result = []
    for sc_members in uf.get_clusters():
        sc_list = sorted(sc_members)
        sc_pairs = {(a, b): s for (a, b), s in pair_scores.items()
                    if a in sc_members and b in sc_members}
        size = len(sc_list)
        conf = compute_cluster_confidence(sc_pairs, size)
        result.append({
            "members": sc_list, "size": size, "oversized": False,
            "pair_scores": sc_pairs, "confidence": conf["confidence"],
            "bottleneck_pair": conf["bottleneck_pair"],
        })
    return result


def build_clusters(
    pairs: list[tuple[int, int, float]],
    all_ids: list[int],
    max_cluster_size: int = 100,
    weak_cluster_threshold: float = 0.3,
) -> dict[int, dict]:
    """Build clusters from scored pairs using Union-Find.

    Auto-splits oversized clusters via MST. Assigns cluster_quality
    ("strong", "weak", "split") and downgrades confidence for weak clusters.
    """
    uf = UnionFind()
    uf.add_many(all_ids)
    for id_a, id_b, _score in pairs:
        uf.union(id_a, id_b)

    clusters = uf.get_clusters()

    member_to_cid: dict[int, int] = {}
    sorted_clusters = sorted(clusters, key=lambda s: min(s))
    for cluster_id, members in enumerate(sorted_clusters, start=1):
        for m in members:
            member_to_cid[m] = cluster_id

    result: dict[int, dict] = {}
    for cluster_id, members in enumerate(sorted_clusters, start=1):
        size = len(members)
        result[cluster_id] = {
            "members": sorted(members),
            "size": size,
            "oversized": size > max_cluster_size,
            "pair_scores": {},
        }

    for id_a, id_b, score in pairs:
        cid = member_to_cid[id_a]
        result[cid]["pair_scores"][(id_a, id_b)] = score

    for cid, cinfo in result.items():
        conf = compute_cluster_confidence(cinfo["pair_scores"], cinfo["size"])
        cinfo["confidence"] = conf["confidence"]
        cinfo["bottleneck_pair"] = conf["bottleneck_pair"]

    # Auto-split oversized clusters
    to_split = [cid for cid, c in result.items() if c["oversized"]]
    while to_split:
        cid = to_split.pop()
        cinfo = result.pop(cid)
        sub_clusters = split_oversized_cluster(cinfo["members"], cinfo["pair_scores"])
        next_cid = max(result.keys(), default=0) + 1
        for sc in sub_clusters:
            sc["oversized"] = sc["size"] > max_cluster_size
            sc["_was_split"] = True
            result[next_cid] = sc
            if sc["oversized"]:
                to_split.append(next_cid)
            next_cid += 1

    # Assign cluster_quality and apply confidence downgrade
    for cid, cinfo in result.items():
        if cinfo.get("_was_split"):
            cinfo["cluster_quality"] = "split"
        elif cinfo["size"] > 1 and cinfo.get("pair_scores"):
            scores = list(cinfo["pair_scores"].values())
            min_edge = min(scores)
            avg_edge = sum(scores) / len(scores)
            if avg_edge - min_edge > weak_cluster_threshold:
                cinfo["cluster_quality"] = "weak"
                cinfo["confidence"] *= 0.7
            else:
                cinfo["cluster_quality"] = "strong"
        else:
            cinfo["cluster_quality"] = "strong"

    return result


def compute_cluster_confidence(
    pair_scores: dict[tuple[int, int], float],
    size: int,
) -> dict:
    """Compute confidence metrics for a cluster.

    Confidence captures how strongly a cluster is connected. A chain
    held together by a single weak link gets low confidence. A fully
    connected cluster with high scores gets high confidence.

    Args:
        pair_scores: Dict of (id_a, id_b) -> score for this cluster.
        size: Number of members in the cluster.

    Returns:
        Dict with: min_edge, avg_edge, connectivity, bottleneck_pair, confidence.
    """
    if size <= 1 or not pair_scores:
        return {
            "min_edge": None,
            "avg_edge": None,
            "connectivity": 1.0 if size <= 1 else 0.0,
            "bottleneck_pair": None,
            "confidence": 1.0 if size <= 1 else 0.0,
        }

    scores = list(pair_scores.values())
    min_edge = min(scores)
    avg_edge = sum(scores) / len(scores)
    max_possible_edges = size * (size - 1) / 2
    connectivity = len(pair_scores) / max_possible_edges if max_possible_edges > 0 else 0.0
    bottleneck_pair = min(pair_scores, key=pair_scores.get)

    # Weighted confidence: weakest link matters most
    confidence = 0.4 * min_edge + 0.3 * avg_edge + 0.3 * connectivity

    return {
        "min_edge": min_edge,
        "avg_edge": avg_edge,
        "connectivity": connectivity,
        "bottleneck_pair": bottleneck_pair,
        "confidence": confidence,
    }


def add_to_cluster(
    record_id: int,
    matches: list[tuple[int, float]],
    clusters: dict[int, dict],
    max_cluster_size: int = 100,
) -> dict[int, dict]:
    """Add a new record to existing clusters based on matches.

    If the record matches members of a single cluster, it joins that cluster.
    If it matches members of multiple clusters, those clusters merge.
    If no matches, a new singleton cluster is created.

    Args:
        record_id: The new record's ID.
        matches: List of (matched_row_id, score) tuples.
        clusters: Current cluster dict (modified in-place).
        max_cluster_size: Threshold for oversized flagging.

    Returns:
        Updated clusters dict.
    """
    if not matches:
        next_cid = max(clusters.keys(), default=0) + 1
        clusters[next_cid] = {
            "members": [record_id],
            "size": 1,
            "oversized": False,
            "pair_scores": {},
            "confidence": 1.0,
            "bottleneck_pair": None,
            "cluster_quality": "strong",
        }
        return clusters

    # Find which cluster(s) the matched records belong to
    member_to_cid: dict[int, int] = {}
    for cid, cinfo in clusters.items():
        for m in cinfo["members"]:
            member_to_cid[m] = cid

    matched_cids = set()
    for matched_id, _score in matches:
        cid = member_to_cid.get(matched_id)
        if cid is not None:
            matched_cids.add(cid)

    if not matched_cids:
        next_cid = max(clusters.keys(), default=0) + 1
        clusters[next_cid] = {
            "members": [record_id],
            "size": 1,
            "oversized": False,
            "pair_scores": {},
            "confidence": 1.0,
            "bottleneck_pair": None,
            "cluster_quality": "strong",
        }
        return clusters

    if len(matched_cids) == 1:
        cid = matched_cids.pop()
        cinfo = clusters[cid]
        cinfo["members"] = sorted(cinfo["members"] + [record_id])
        cinfo["size"] += 1
        cinfo["oversized"] = cinfo["size"] > max_cluster_size
        for matched_id, score in matches:
            if member_to_cid.get(matched_id) == cid:
                cinfo["pair_scores"][(min(record_id, matched_id), max(record_id, matched_id))] = score
        conf = compute_cluster_confidence(cinfo["pair_scores"], cinfo["size"])
        cinfo["confidence"] = conf["confidence"]
        cinfo["bottleneck_pair"] = conf["bottleneck_pair"]
        cinfo["cluster_quality"] = cinfo.get("cluster_quality", "strong")
        return clusters

    # Multiple clusters — merge them all with the new record
    merged_members = [record_id]
    merged_pairs: dict[tuple[int, int], float] = {}

    for cid in matched_cids:
        cinfo = clusters[cid]
        merged_members.extend(cinfo["members"])
        merged_pairs.update(cinfo["pair_scores"])
        del clusters[cid]

    # Add new pair scores
    for matched_id, score in matches:
        merged_pairs[(min(record_id, matched_id), max(record_id, matched_id))] = score

    next_cid = max(clusters.keys(), default=0) + 1
    size = len(merged_members)
    conf = compute_cluster_confidence(merged_pairs, size)
    clusters[next_cid] = {
        "members": sorted(merged_members),
        "size": size,
        "oversized": size > max_cluster_size,
        "pair_scores": merged_pairs,
        "confidence": conf["confidence"],
        "bottleneck_pair": conf["bottleneck_pair"],
        "cluster_quality": "strong",
    }

    return clusters


def get_cluster_pair_scores(
    cluster_members: list[int],
    all_pairs: list[tuple[int, int, float]],
) -> dict[tuple[int, int], float]:
    """Get pair scores for a specific cluster. Call on-demand, not in hot path."""
    member_set = set(cluster_members)
    return {
        (a, b): s
        for a, b, s in all_pairs
        if a in member_set and b in member_set
    }


def unmerge_record(
    record_id: int,
    clusters: dict[int, dict],
    threshold: float = 0.0,
) -> dict[int, dict]:
    """Remove a record from its cluster and re-cluster remaining members.

    Uses stored pair_scores to re-build the cluster without the removed record.
    The removed record becomes a singleton. If the remaining members still form
    connected components above the threshold, they stay clustered.

    Args:
        record_id: The record ID to remove from its cluster.
        clusters: Full cluster dict (modified in-place and returned).
        threshold: Minimum score to keep a pair connection (default 0.0 = keep all).

    Returns:
        Updated clusters dict. The removed record is in its own singleton cluster.
    """
    # Find which cluster contains this record
    source_cid = None
    for cid, cinfo in clusters.items():
        if record_id in cinfo["members"]:
            source_cid = cid
            break

    if source_cid is None:
        return clusters  # Record not found in any cluster

    cinfo = clusters[source_cid]
    if cinfo["size"] <= 1:
        return clusters  # Already a singleton

    # Extract pair_scores excluding the removed record
    remaining_members = [m for m in cinfo["members"] if m != record_id]
    remaining_pairs = [
        (a, b, s)
        for (a, b), s in cinfo["pair_scores"].items()
        if a != record_id and b != record_id and s >= threshold
    ]

    # Re-cluster the remaining members
    sub_clusters = build_clusters(remaining_pairs, remaining_members)

    # Remove the original cluster
    del clusters[source_cid]

    # Assign new cluster IDs (use max existing + 1)
    next_cid = max(clusters.keys(), default=0) + 1

    # Add the removed record as a singleton
    clusters[next_cid] = {
        "members": [record_id],
        "size": 1,
        "oversized": False,
        "pair_scores": {},
    }
    next_cid += 1

    # Add re-clustered groups
    for sub_cinfo in sub_clusters.values():
        clusters[next_cid] = sub_cinfo
        next_cid += 1

    return clusters


def unmerge_cluster(
    cluster_id: int,
    clusters: dict[int, dict],
) -> dict[int, dict]:
    """Shatter a cluster back into individual singletons.

    All members become their own cluster. Pair scores are discarded.

    Args:
        cluster_id: The cluster to shatter.
        clusters: Full cluster dict (modified in-place and returned).

    Returns:
        Updated clusters dict with the cluster replaced by singletons.
    """
    if cluster_id not in clusters:
        return clusters

    cinfo = clusters[cluster_id]
    members = cinfo["members"]
    del clusters[cluster_id]

    next_cid = max(clusters.keys(), default=0) + 1
    for member_id in members:
        clusters[next_cid] = {
            "members": [member_id],
            "size": 1,
            "oversized": False,
            "pair_scores": {},
        }
        next_cid += 1

    return clusters

File: packages/goldenmatch-js/src/core/cluster.ts
/**
 * cluster.ts — Union-Find clustering with MST splitting.
 * Edge-safe: no Node.js imports, pure TypeScript only.
 */

import type { ClusterInfo } from "./types.js";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Canonicalize a pair key: always min:max. */
export function pairKey(a: number, b: number): string {
  return `${Math.min(a, b)}:${Math.max(a, b)}`;
}

/** Parse a pair key back into [idA, idB]. */
function parsePairKey(key: string): [number, number] {
  const idx = key.indexOf(":");
  return [Number(key.slice(0, idx)), Number(key.slice(idx + 1))];
}

// ---------------------------------------------------------------------------
// UnionFind
// ---------------------------------------------------------------------------

export class UnionFind {
  private parent = new Map<number, number>();
  private rank = new Map<number, number>();

  /** Add element as its own root. */
  add(x: number): void {
    if (!this.parent.has(x)) {
      this.parent.set(x, x);
      this.rank.set(x, 0);
    }
  }

  /** Batch add multiple elements. */
  addMany(ids: readonly number[]): void {
    for (const x of ids) {
      if (!this.parent.has(x)) {
        this.parent.set(x, x);
        this.rank.set(x, 0);
      }
    }
  }

  /** Find root with iterative path compression. */
  find(x: number): number {
    let root = x;
    while (this.parent.get(root) !== root) {
      root = this.parent.get(root)!;
    }
    // Path compression
    let current = x;
    while (this.parent.get(current) !== root) {
      const next = this.parent.get(current)!;
      this.parent.set(current, root);
      current = next;
    }
    return root;
  }

  /** Union by rank. */
  union(a: number, b: number): void {
    let ra = this.find(a);
    let rb = this.find(b);
    if (ra === rb) return;
    const rankA = this.rank.get(ra)!;
    const rankB = this.rank.get(rb)!;
    if (rankA < rankB) {
      [ra, rb] = [rb, ra];
    }
    this.parent.set(rb, ra);
    if (rankA === rankB) {
      this.rank.set(ra, rankA + 1);
    }
  }

  /** Return all clusters as arrays of sets. */
  getClusters(): Set<number>[] {
    const groups = new Map<number, Set<number>>();
    for (const x of this.parent.keys()) {
      const root = this.find(x);
      let group = groups.get(root);
      if (!group) {
        group = new Set<number>();
        groups.set(root, group);
      }
      group.add(x);
    }
    return Array.from(groups.values());
  }
}

// ---------------------------------------------------------------------------
// MST (max-weight spanning tree via Kruskal)
// ---------------------------------------------------------------------------

/**
 * Build a max-weight spanning tree using Kruskal's algorithm.
 * Returns edges as [idA, idB, score] sorted by descending weight.
 */
export function buildMst(
  members: readonly number[],
  pairScores: ReadonlyMap<string, number>,
): [number, number, number][] {
  // Collect and sort edges by score descending
  const edges: [number, number, number][] = [];
  for (const [key, score] of pairScores) {
    const [a, b] = parsePairKey(key);
    edges.push([a, b, score]);
  }
  edges.sort((x, y) => y[2] - x[2]);

  const uf = new UnionFind();
  uf.addMany(members);

  const mst: [number, number, number][] = [];
  const target = members.length - 1;
  for (const [a, b, s] of edges) {
    if (uf.find(a) !== uf.find(b)) {
      uf.union(a, b);
      mst.push([a, b, s]);
      if (mst.length === target) break;
    }
  }
  return mst;
}

// ---------------------------------------------------------------------------
// Cluster confidence
// ---------------------------------------------------------------------------

export interface ClusterConfidence {
  readonly minEdge: number | null;
  readonly avgEdge: number | null;
  readonly connectivity: number;
  readonly bottleneckPair: readonly [number, number] | null;
  readonly confidence: number;
}

/**
 * Compute confidence metrics for a cluster.
 * confidence = 0.4 * minEdge + 0.3 * avgEdge + 0.3 * connectivity
 */
export function computeClusterConfidence(
  pairScores: ReadonlyMap<string, number>,
  size: number,
): ClusterConfidence {
  if (size <= 1 || pairScores.size === 0) {
    return {
      minEdge: null,
      avgEdge: null,
      connectivity: size <= 1 ? 1.0 : 0.0,
      bottleneckPair: null,
      confidence: size <= 1 ? 1.0 : 0.0,
    };
  }

  let minEdge = Infinity;
  let sum = 0;
  let bottleneckKey = "";

  for (const [key, score] of pairScores) {
    sum += score;
    if (score < minEdge) {
      minEdge = score;
      bottleneckKey = key;
    }
  }

  const avgEdge = sum / pairScores.size;
  const maxPossibleEdges = (size * (size - 1)) / 2;
  const connectivity =
    maxPossibleEdges > 0 ? pairScores.size / maxPossibleEdges : 0.0;

  const bottleneckPair: readonly [number, number] | null = bottleneckKey
    ? parsePairKey(bottleneckKey)
    : null;

  const confidence = 0.4 * minEdge + 0.3 * avgEdge + 0.3 * connectivity;

  return { minEdge, avgEdge, connectivity, bottleneckPair, confidence };
}

// ---------------------------------------------------------------------------
// Split oversized cluster
// ---------------------------------------------------------------------------

/** Internal mutable cluster info used during building. */
interface MutableClusterInfo {
  members: number[];
  size: number;
  oversized: boolean;
  pairScores: Map<string, number>;
  confidence: number;
  bottleneckPair: readonly [number, number] | null;
  clusterQuality: "strong" | "weak" | "split";
  _wasSplit?: boolean;
}

/**
 * Split a cluster by removing the weakest MST edge.
 * Returns sub-cluster infos.
 */
export function splitOversizedCluster(
  members: readonly number[],
  pairScores: ReadonlyMap<string, number>,
): MutableClusterInfo[] {
  if (members.length <= 1 || pairScores.size === 0) {
    return [
      {
        members: [...members].sort((a, b) => a - b),
        size: members.length,
        oversized: false,
        pairScores: new Map(pairScores),
        confidence: 1.0,
        bottleneckPair: null,
        clusterQuality: "strong",
      },
    ];
  }

  const mst = buildMst(members, pairScores);
  if (mst.length === 0) {
    return [
      {
        members: [...members].sort((a, b) => a - b),
        size: members.length,
        oversized: false,
        pairScores: new Map(pairScores),
        confidence: 1.0,
        bottleneckPair: null,
        clusterQuality: "strong",
      },
    ];
  }

  // Find weakest edge
  let weakestIdx = 0;
  let weakestScore = mst[0]![2];
  for (let i = 1; i < mst.length; i++) {
    if (mst[i]![2] < weakestScore) {
      weakestScore = mst[i]![2];
      weakestIdx = i;
    }
  }

  // Rebuild without weakest edge
  const uf = new UnionFind();
  uf.addMany(members as number[]);
  for (let i = 0; i < mst.length; i++) {
    if (i !== weakestIdx) {
      uf.union(mst[i]![0], mst[i]![1]);
    }
  }

  const result: MutableClusterInfo[] = [];
  for (const subMembers of uf.getClusters()) {
    const subList = [...subMembers].sort((a, b) => a - b);
    const subPairs = new Map<string, number>();
    for (const [key, score] of pairScores) {
      const [a, b] = parsePairKey(key);
      if (subMembers.has(a) && subMembers.has(b)) {
        subPairs.set(key, score);
      }
    }
    const conf = computeClusterConfidence(subPairs, subList.length);
    result.push({
      members: subList,
      size: subList.length,
      oversized: false,
      pairScores: subPairs,
      confidence: conf.confidence,
      bottleneckPair: conf.bottleneckPair,
      clusterQuality: "strong",
    });
  }
  return result;
}

// ---------------------------------------------------------------------------
// buildClusters options
// ---------------------------------------------------------------------------

export interface BuildClustersOptions {
  readonly maxClusterSize?: number;
  readonly weakClusterThreshold?: number;
  readonly autoSplit?: boolean;
}

// ---------------------------------------------------------------------------
// buildClusters
// ---------------------------------------------------------------------------

/**
 * Build clusters from scored pairs using Union-Find.
 *
 * Auto-splits oversized clusters via MST (iterative, not recursive).
 * Assigns cluster_quality: "strong", "weak" (avg-min > weakThreshold), or "split".
 * Downgrades confidence by 0.7 for weak clusters.
 */
export function buildClusters(
  pairs: readonly (readonly [number, number, number])[],
  allIds: readonly number[],
  options?: BuildClustersOptions,
): Map<number, ClusterInfo> {
  const maxClusterSize = options?.maxClusterSize ?? 100;
  const weakClusterThreshold = options?.weakClusterThreshold ?? 0.3;
  const autoSplit = options?.autoSplit ?? true;

  // Build Union-Find from pairs
  const uf = new UnionFind();
  uf.addMany(allIds);
  for (const [idA, idB] of pairs) {
    uf.union(idA, idB);
  }

  const clusters = uf.getClusters();

  // Sort clusters by minimum member for deterministic IDs.
  // Use for-loop min — Math.min(...set) crashes on Sets with >65K elements.
  const minOf = (s: Set<number>): number => {
    let m = Infinity;
    for (const v of s) if (v < m) m = v;
    return m;
  };
  clusters.sort((a, b) => minOf(a) - minOf(b));

  // Map members to cluster IDs
  const memberToCid = new Map<number, number>();
  for (let i = 0; i < clusters.length; i++) {
    const cid = i + 1;
    for (const m of clusters[i]!) {
      memberToCid.set(m, cid);
    }
  }

  // Build mutable result
  const result = new Map<number, MutableClusterInfo>();
  for (let i = 0; i < clusters.length; i++) {
    const cid = i + 1;
    const memberArr = [...clusters[i]!].sort((a, b) => a - b);
    result.set(cid, {
      members: memberArr,
      size: memberArr.length,
      oversized: memberArr.length > maxClusterSize,
      pairScores: new Map(),
      confidence: 0,
      bottleneckPair: null,
      clusterQuality: "strong",
    });
  }

  // Assign pair scores to clusters (canonicalized keys)
  for (const [idA, idB, score] of pairs) {
    const cid = memberToCid.get(idA)!;
    const info = result.get(cid)!;
    info.pairScores.set(pairKey(idA, idB), score);
  }

  // Compute initial confidence
  for (const [, cinfo] of result) {
    const conf = computeClusterConfidence(cinfo.pairScores, cinfo.size);
    cinfo.confidence = conf.confidence;
    cinfo.bottleneckPair = conf.bottleneckPair;
  }

  // Auto-split oversized clusters (iterative)
  if (autoSplit) {
    const toSplit: number[] = [];
    for (const [cid, c] of result) {
      if (c.oversized) toSplit.push(cid);
    }

    while (toSplit.length > 0) {
      const cid = toSplit.pop()!;
      const cinfo = result.get(cid)!;
      result.delete(cid);

      const subClusters = splitOversizedCluster(
        cinfo.members,
        cinfo.pairScores,
      );
      let nextCid = 0;
      for (const [k] of result) {
        if (k > nextCid) nextCid = k;
      }
      nextCid += 1;

      for (const sc of subClusters) {
        sc.oversized = sc.size > maxClusterSize;
        sc._wasSplit = true;
        result.set(nextCid, sc);
        if (sc.oversized) {
          toSplit.push(nextCid);
        }
        nextCid++;
      }
    }
  }

  // Assign cluster_quality and apply confidence downgrade
  for (const [, cinfo] of result) {
    if (cinfo._wasSplit) {
      cinfo.clusterQuality = "split";
    } else if (cinfo.size > 1 && cinfo.pairScores.size > 0) {
      const scores = [...cinfo.pairScores.values()];
      let minE = Infinity;
      let sumE = 0;
      for (const s of scores) {
        if (s < minE) minE = s;
        sumE += s;
      }
      const avgE = sumE / scores.length;
      if (avgE - minE > weakClusterThreshold) {
        cinfo.clusterQuality = "weak";
        cinfo.confidence *= 0.7;
      } else {
        cinfo.clusterQuality = "strong";
      }
    } else {
      cinfo.clusterQuality = "strong";
    }
    delete cinfo._wasSplit;
  }

  // Freeze into ClusterInfo
  const frozen = new Map<number, ClusterInfo>();
  for (const [cid, c] of result) {
    frozen.set(cid, {
      members: c.members,
      size: c.size,
      oversized: c.oversized,
      pairScores: c.pairScores,
      confidence: c.confidence,
      bottleneckPair: c.bottleneckPair,
      clusterQuality: c.clusterQuality,
    });
  }
  return frozen;
}

// ---------------------------------------------------------------------------
// addToCluster
// ---------------------------------------------------------------------------

/**
 * Add a new record to existing clusters based on matches.
 *
 * - No matches: new singleton cluster
 * - Single cluster match: join that cluster
 * - Multiple cluster match: merge all matched clusters
 *
 * Flags oversized but does NOT auto-split. Caller should call
 * splitOversizedCluster() if desired.
 */
export function addToCluster(
  recordId: number,
  matches: readonly (readonly [number, number])[],
  clusters: Map<number, ClusterInfo>,
  maxClusterSize = 100,
): Map<number, ClusterInfo> {
  const makeSingleton = (): ClusterInfo => ({
    members: [recordId],
    size: 1,
    oversized: false,
    pairScores: new Map(),
    confidence: 1.0,
    bottleneckPair: null,
    clusterQuality: "strong",
  });

  if (matches.length === 0) {
    const nextCid = _nextCid(clusters);
    clusters.set(nextCid, makeSingleton());
    return clusters;
  }

  // Map members to cluster IDs
  const memberToCid = new Map<number, number>();
  for (const [cid, cinfo] of clusters) {
    for (const m of cinfo.members) {
      memberToCid.set(m, cid);
    }
  }

  const matchedCids = new Set<number>();
  for (const [matchedId] of matches) {
    const cid = memberToCid.get(matchedId);
    if (cid !== undefined) matchedCids.add(cid);
  }

  if (matchedCids.size === 0) {
    const nextCid = _nextCid(clusters);
    clusters.set(nextCid, makeSingleton());
    return clusters;
  }

  if (matchedCids.size === 1) {
    const cid = matchedCids.values().next().value!;
    const old = clusters.get(cid)!;
    const newPairs = new Map(old.pairScores);

    for (const [matchedId, score] of matches) {
      if (memberToCid.get(matchedId) === cid) {
        newPairs.set(pairKey(recordId, matchedId), score);
      }
    }

    const newMembers = [...old.members, recordId].sort((a, b) => a - b);
    const newSize = newMembers.length;
    const conf = computeClusterConfidence(newPairs, newSize);

    clusters.set(cid, {
      members: newMembers,
      size: newSize,
      oversized: newSize > maxClusterSize,
      pairScores: newPairs,
      confidence: conf.confidence,
      bottleneckPair: conf.bottleneckPair,
      clusterQuality: old.clusterQuality,
    });
    return clusters;
  }

  // Multiple clusters: merge all
  const mergedMembers: number[] = [recordId];
  const mergedPairs = new Map<string, number>();

  for (const cid of matchedCids) {
    const cinfo = clusters.get(cid)!;
    mergedMembers.push(...cinfo.members);
    for (const [k, v] of cinfo.pairScores) {
      mergedPairs.set(k, v);
    }
    clusters.delete(cid);
  }

  for (const [matchedId, score] of matches) {
    mergedPairs.set(pairKey(recordId, matchedId), score);
  }

  const sortedMembers = mergedMembers.sort((a, b) => a - b);
  const size = sortedMembers.length;
  const conf = computeClusterConfidence(mergedPairs, size);
  const nextCid = _nextCid(clusters);

  clusters.set(nextCid, {
    members: sortedMembers,
    size,
    oversized: size > maxClusterSize,
    pairScores: mergedPairs,
    confidence: conf.confidence,
    bottleneckPair: conf.bottleneckPair,
    clusterQuality: "strong",
  });

  return clusters;
}

// ---------------------------------------------------------------------------
// unmergeRecord
// ---------------------------------------------------------------------------

/**
 * Remove a record from its cluster and re-cluster remaining members.
 * The removed record becomes a singleton.
 */
export function unmergeRecord(
  recordId: number,
  clusters: Map<number, ClusterInfo>,
  threshold = 0.0,
): Map<number, ClusterInfo> {
  // Find which cluster contains this record
  let sourceCid: number | null = null;
  for (const [cid, cinfo] of clusters) {
    if (cinfo.members.includes(recordId)) {
      sourceCid = cid;
      break;
    }
  }

  if (sourceCid === null) return clusters; // Not found
  const cinfo = clusters.get(sourceCid)!;
  if (cinfo.size <= 1) return clusters; // Already singleton

  // Extract pairs excluding the removed record, applying threshold
  const remainingMembers = cinfo.members.filter((m) => m !== recordId);
  const remainingPairs: [number, number, number][] = [];
  for (const [key, score] of cinfo.pairScores) {
    const [a, b] = parsePairKey(key);
    if (a !== recordId && b !== recordId && score >= threshold) {
      remainingPairs.push([a, b, score]);
    }
  }

  // Re-cluster remaining members
  const subClusters = buildClusters(remainingPairs, remainingMembers);

  // Remove the original cluster
  clusters.delete(sourceCid);

  // Assign new cluster IDs
  let nextCid = _nextCid(clusters);

  // Add the removed record as a singleton
  clusters.set(nextCid, {
    members: [recordId],
    size: 1,
    oversized: false,
    pairScores: new Map(),
    confidence: 1.0,
    bottleneckPair: null,
    clusterQuality: "strong",
  });
  nextCid++;

  // Add re-clustered groups
  for (const [, subInfo] of subClusters) {
    clusters.set(nextCid, subInfo);
    nextCid++;
  }

  return clusters;
}

// ---------------------------------------------------------------------------
// unmergeCluster
// ---------------------------------------------------------------------------

/**
 * Shatter a cluster into individual singletons.
 * All members become their own cluster. Pair scores are discarded.
 */
export function unmergeCluster(
  clusterId: number,
  clusters: Map<number, ClusterInfo>,
): Map<number, ClusterInfo> {
  const cinfo = clusters.get(clusterId);
  if (!cinfo) return clusters;

  const members = cinfo.members;
  clusters.delete(clusterId);

  let nextCid = _nextCid(clusters);
  for (const memberId of members) {
    clusters.set(nextCid, {
      members: [memberId],
      size: 1,
      oversized: false,
      pairScores: new Map(),
      confidence: 1.0,
      bottleneckPair: null,
      clusterQuality: "strong",
    });
    nextCid++;
  }

  return clusters;
}

// ---------------------------------------------------------------------------
// getClusterPairScores
// ---------------------------------------------------------------------------

/**
 * Get pair scores for a specific set of cluster members from all pairs.
 * Call on-demand, not in hot path.
 */
export function getClusterPairScores(
  members: readonly number[],
  allPairs: readonly (readonly [number, number, number])[],
): Map<string, number> {
  const memberSet = new Set(members);
  const result = new Map<string, number>();
  for (const [a, b, s] of allPairs) {
    if (memberSet.has(a) && memberSet.has(b)) {
      result.set(pairKey(a, b), s);
    }
  }
  return result;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function _nextCid(clusters: ReadonlyMap<number, unknown>): number {
  let max = 0;
  for (const k of clusters.keys()) {
    if (k > max) max = k;
  }
  return max + 1;
}
